### PR TITLE
User qualifiers

### PR DESCRIPTION
**Pull request by atgeller**
Add support for user defined qualifiers which get checked for well formedness and passed along to liquid-fixpoint

### CODE REVIEW

File: liquid-rust-fixpoint/src/lib.rs [hunk 1]
```diff
@@ -91,6 +92,10 @@ impl<Tag: fmt::Display> fmt::Display for Task<Tag> {
             writeln!(f, "{qualif}")?;
         }
 
+        for qualif in self.qualifiers.iter() {
```
Comment (nilehmann) [inline liquid-rust-fixpoint/src/lib.rs:95]: I typically prefer
```suggestion
        for qualif in &self.qualifiers {
```

File: liquid-rust-syntax/src/ast.rs [hunk 2]
```diff
@@ -37,6 +37,14 @@ pub struct FnSig {
     pub span: Span,
 }
 
+#[derive(Debug)]
+pub struct Qualifier {
+    pub name: Ident,
+    pub args: Vec<(Ident, Ident)>,
```
Comment (nilehmann) [inline liquid-rust-syntax/src/ast.rs:43]: Define a struct for `(Ident, Ident)` there's no way to know which one is which by just looking at the types. Maybe you can call it `QualifParam` or something

File: liquid-rust-syntax/src/grammar.lalrpop [hunk 3]
```diff
@@ -192,6 +203,7 @@ extern {
     enum Token {
         "literal" => Token::Literal(<Lit>),
         "ident"   => Token::Ident(<Symbol>),
+        "qualifier" => Token::Qualifier,
```
Comment (nilehmann) [inline liquid-rust-syntax/src/grammar.lalrpop:206]: this is not being used
```suggestion
```

File: liquid-rust-syntax/src/lexer.rs [hunk 4]
```diff
@@ -38,6 +38,7 @@ pub enum Token {
     And,
     Percent,
     Weak,
+    Qualifier,
```
Comment (nilehmann) [inline liquid-rust-syntax/src/lexer.rs:41]: not being used

```suggestion
```

File: liquid-rust-tests/tests/neg/user_qualifiers/not_well_formed.rs [hunk 5]
```diff
@@ -0,0 +1,11 @@
+#![feature(register_tool)]
+#![register_tool(lr)]
+#![feature(custom_inner_attributes)]
+
+#![lr::qualifier(MyQ (x: int) -> x == a)] //~ ERROR cannot find value `a` in this scope
+#![lr::qualifier(MyQ (x: int) -> x + 1)] //~ ERROR mismatched sorts
```
Comment (nilehmann) [inline liquid-rust-tests/tests/neg/user_qualifiers/not_well_formed.rs:6]: I kind of hate these spaces 😛 

```suggestion
#![lr::qualifier(MyQ(x: int) -> x == a)] //~ ERROR cannot find value `a` in this scope
#![lr::qualifier(MyQ(x: int) -> x + 1)] //~ ERROR mismatched sorts
```

File: liquid-rust-tests/tests/pos/user_qualifiers/well_formed.rs [hunk 6]
```diff
@@ -0,0 +1,11 @@
+#![feature(register_tool)]
+#![register_tool(lr)]
+#![feature(custom_inner_attributes)]
+
+#![lr::qualifier(MyQ1 (x: int, a: int) -> x == a)]
+#![lr::qualifier(MyQ2 (x: int) -> x > 1)]
```
Comment (nilehmann) [inline liquid-rust-tests/tests/pos/user_qualifiers/well_formed.rs:6]:
```suggestion
#![lr::qualifier(MyQ1(x: int, a: int) -> x == a)]
#![lr::qualifier(MyQ2(x: int) -> x > 1)]
```

File: liquid-rust-typeck/src/lib.rs [hunk 7]
```diff
@@ -76,7 +77,12 @@ pub fn check<'tcx>(
 
     let constraint = pure_cx.into_fixpoint(&mut fcx);
 
-    fcx.check(genv.tcx, def_id, body.mir.span, constraint)
+    let qualifiers = qualifiers
```
Comment (nilehmann) [inline liquid-rust-typeck/src/lib.rs:80]: Lower this in `callbacks.rs`

File: liquid-rust-typeck/src/lib.rs [hunk 8]
```diff
@@ -112,9 +118,17 @@ impl FixpointCtxt {
...
+        qualifiers: Vec<ty::Qualifier>,
     ) -> Result<(), ErrorReported> {
         let kvars = self.kvars.into_fixpoint();
-        let task = fixpoint::Task::new(kvars, constraint);
+
+        let mut fx_qualifiers = Vec::new();
+
+        for qualifier in qualifiers {
+            fx_qualifiers.push(qualifier_into_fixpoint(&qualifier));
+        }
```
Comment (nilehmann) [inline liquid-rust-typeck/src/lib.rs:129]: Prefer a functional style, this allows you to reuse names.

```suggestion
        let qualifiers = qualifier.iter().map(qualifier_into_fixpoint).collect()
```

File: liquid-rust-typeck/src/wf.rs [hunk 9]
```diff
@@ -70,6 +70,14 @@ impl Wf<'_> {
         Ok(())
     }
 
+    pub fn check_qualifier(&self, qualifier: &core::Qualifier) -> Result<(), ErrorReported> {
+        let mut env = Env::new(&qualifier.args);
+
+        let _ = self.check_expr(&mut env, &qualifier.expr, ty::Sort::bool())?;
+
+        Ok(())
```
Comment (nilehmann) [inline liquid-rust-typeck/src/wf.rs:78]:
```suggestion
        self.check_expr(&mut env, &qualifier.expr, ty::Sort::bool())
```

File: liquid-rust-driver/src/callbacks.rs [hunk 10]
```diff
@@ -42,6 +42,17 @@ fn check_crate(tcx: TyCtxt, sess: &Session) -> Result<(), ErrorReported> {
         .try_collect_exhaust()?;
 
     let wf = Wf::new(sess, &adt_defs);
+
+    let qualifiers: Vec<liquid_rust_core::ty::Qualifier> = specs
+        .qualifs
+        .into_iter()
+        .map(|qualifier| {
+            let resolved = Resolver::resolve_qualifier(tcx, qualifier)?;
```
Comment (nilehmann) [inline liquid-rust-driver/src/callbacks.rs:50]: you should add the `lowering` here

**Review by nilehmann — APPROVED** [review]
@atgeller Looking great! just some nitpicking. Feel free to merge after you've gone over my comments.

File: liquid-rust-typeck/src/lib.rs [hunk 11]
```diff
@@ -76,7 +77,12 @@ pub fn check<'tcx>(
 
     let constraint = pure_cx.into_fixpoint(&mut fcx);
 
-    fcx.check(genv.tcx, def_id, body.mir.span, constraint)
+    let qualifiers = qualifiers
+        .iter()
+        .map(|q| -> ty::Qualifier { lowering::LoweringCtxt::lower_qualifer(q) })
```
Comment (nilehmann) [inline liquid-rust-typeck/src/lib.rs:82]: I didn't know this syntax was valid. Can you omit the `->`?

```suggestion
        .map(|q| ty::Qualifier { lowering::LoweringCtxt::lower_qualifer(q) })
```

Comment (atgeller) [reply liquid-rust-typeck/src/lib.rs:82]: The whole `map` there goes away when we move to lowering in callbacks.

**Comment by atgeller** [comment]
All changes made. I'll go ahead and merge if it is mergable

**Pull request merged**